lambda/matching.ml

```diff
@@ -3773,38 +3844,26 @@ let failure_handler ~scopes loc ~failer () =
         ],
         sloc )
 
-let check_total ~scopes loc ~failer total lambda i =
-  if Jumps.is_empty total then
-    lambda
-  else
-    Lstaticcatch (lambda, (i, []),
-                  failure_handler ~scopes loc ~failer ())
-
 let toplevel_handler ~scopes loc ~failer partial args cases compile_fun =
   let compile_fun partial pm =
     debugf "@[<v>MATCHING@,";
     let result = compile_fun partial pm in
     debugf "@]@.";
     result
   in
-  match partial with
-  | Total when not !Clflags.safer_matching ->
-      let default = Default_environment.empty in
-      let pm = { args; cases; default } in
-      let (lam, total) = compile_fun Total pm in
-      assert (Jumps.is_empty total);
-      lam
-  | Partial | Total (* when !Clflags.safer_matching *) ->
-      let raise_num = next_raise_count () in
-      let default =
-        Default_environment.cons [ Patterns.omega_list args ] raise_num
-          Default_environment.empty in
-      let pm = { args; cases; default } in
-      begin match compile_fun Partial pm with
-      | exception Unused -> assert false
-      | (lam, total) ->
-          check_total ~scopes loc ~failer total lam raise_num
-      end
+  let final_exit = next_raise_count () in
+  let default = Default_environment.empty ~final_exit in
+  let pm = { args; cases; default } in
+  let safe_partial = if !Clflags.safer_matching then Partial else partial in
+  begin match compile_fun safe_partial pm with
+  | exception Unused -> assert false
+  | (lam, jumps) ->
+      match Jumps.partial jumps with
```

Comment:
There is a global invariant that functions that take a `default` environment return a jump summary that is well-scoped with respect to this default environment, it only mentions exit numbers that are listed in the default environment. This is more important than specifically ensuring an empty environment here -- it holds in recursive calls as well.

I thought about how to check this statically upon reading your comment. My intuition is that it holds "by construction" because all extensions of the jump summary are done on exit numbers obtained from the default environment. In fact when reading the code I noticed that the present PR breaks this invariant subtle in the `mk_failaction_pos` case, it sometimes calls `Jumps.add i ... jumps` when `i` is the final exit, and not one the non-final exits explicitly listed in the default environment. I changed this by pushing an extra commit which you can have a look at if you are curious.

(The shape of the code is slightly tricky here because we accumulate exit numbers and fail patterns to build both the `fails` list and the jump summary, but we don't want to handle the final exit differently for fail patterns.)

To summarize. (1) in fact the idea that `jumps.env` is empty at this point *did not hold* with the code that you looked at, but (2) this was an innocuous difference because mentioning the final exit in the jump summaries does not make a difference as long as the "partiality" information of the jump summary is correct, and (3) this should be fixed now.